### PR TITLE
Edit types to parameterize value. Edit validations rules to apply to specific types

### PR DESCRIPTION
**Pull request by jlefebure**
See #10

### CODE REVIEW

**Comment by rkuykendall** [comment]
Thank you so much @jlefebure! This is really fantastic work. I don't have much experience with parameterization and could never have done this on my own. Before merge and release though, I'd like to slim this PR down to *just* the type changes, so if you could do a few things for me:

1. Review this PR to remove the circular import from utils: https://github.com/formsy/formsy-react/pull/362 (other changes in the description)
2. Revert all changes in `dist`. I know it's useful to build while working, but all our dist changes are handled by the release script now, which makes it much easier to review PRs.
3. Revert the changes to `yarn.lock` unless necessary for types. I ran and merged automatic dependency bumps this morning so it should be fine if you needed something updated.

**Comment by rkuykendall** [comment]
Also, if you are very busy and would prefer, we can set this PR to merge into a new branch off master and I'll do that work. You've already contributed the hard part. I just want to make sure you get as much credit as possible so I'd love to have your name on the commit :)

**Comment by jlefebure** [comment]
Sure, I would love to. #362 seems to add some utility methods that I can use in the validation functions. I already push points 2 and 3. When #362 will be merged, I will rebase this PR on master and make the necessary ajustements (as well as using utility methods)

**Comment by rkuykendall** [comment]
@jlefebure Merged! It also seems we have some repetition with our `if string, val === '', else val === undefined` logic so maybe we should make that a util? like `isEmptyValue`? But that's easily something we could push off until post-2.x

**Comment by rkuykendall** [comment]
I re-read the PR this morning just to make sure I wasn't missing anything, but the types look great :) Excited for the rebase changes.

**Comment by jlefebure** [comment]
Hello, 

All done ! Rebased from master, with value type and utils methods. Tests and lint are OK. I have included some more utils functions and using them into the validation rules. I don't think there is anything else from my side, but don't hesitate to tell me for any other ajustement.

File: src/utils.ts [hunk 1]
```diff
@@ -8,6 +8,10 @@ export function isObject(value: unknown): value is object {
   return value !== null && typeof value === 'object';
 }
 
+export function isTypeUndefined(value: unknown): value is object {
```
Comment (rkuykendall) [inline src/utils.ts]:
```suggestion
export function isTypeUndefined(value: unknown): value is undefined {
```
^ Copy-paste error I think?

Comment (rkuykendall) [reply src/utils.ts]: @jlefebure Commit this if I'm right, or just let me know if I'm missing something. Then I'll merge!! :D

Comment (jlefebure) [reply src/utils.ts]: Woops, yes it is

**Comment by rkuykendall** [comment]
Done! https://github.com/formsy/formsy-react/releases/tag/v2.0.0-beta.8

**Pull request merged**